### PR TITLE
UX: add information about missing optional dependencies as warnings and error messages in yt.load

### LINKED ISSUE #4274: UX: add hints to YTUnidentifiedDataType
### Bug report

**Bug summary**

`YTUnidentifiedDataType` is the error ones get from `yt.load(...)` when no "valid" frontend is found. Currently the error message  doesn't give much guidance, but in many a case it can be fixed by either:
- installing missing requirements (most often `h5py`), because many implementations of `Dataset._is_valid()` just return `False` when getting an import error. 
- upgrading yt (if the current version is older than the first one to have the required frontend)

It would be helpful to acknowledge these solutions in the message. Ideally, I think we should avoid importing stuff in `_is_valid` methods, but that calls for a broad refactor with a high chance of breaking someone's workflow, so I think it's a little late to do that.

I note that #4272 is helpful here because it gathers the important data (frontend VS requirements mapping) in one place (`pyproject.toml`).

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary
Fix #4274
For now I've only tested this manually but it should be relatively straightforward to add tests.

TODO:
- [x] tests
- [x] clean up redundant checks and warnings (`warn_h5py` and the likes) from `_is_valid` methods
- [x] implement early returns in `_is_valid` methods when requirements are missing
- [x] rebase

### CODE REVIEW

**Review by chrishavlin — COMMENTED** [review]
I'd like to check out the branch and play around with this before hitting the approve button, but here are a few comments and questions in the mean time. Generally looks good to me though!

File: yt/data_objects/static_output.py
Comment (chrishavlin) [reply yt/data_objects/static_output.py:399]: context to my question: yt.load accepts `PathLike` objects, does that get coerced into a str before `_is_valid` is called?

Comment (chrishavlin) [reply yt/data_objects/static_output.py:399]: oh nevermind I just looked...

Comment (chrishavlin) [reply yt/data_objects/static_output.py:399]: and I think it does

Comment (chrishavlin) [reply yt/data_objects/static_output.py:399]: Ok, ya, ` os.path.expanduser()` will always return a string even if the `filename` is a `Path`, soooo I'll resolve this conversation with myself now :)

File: yt/frontends/chombo/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/chombo/data_structures.py:643]: I had a preference for overloading the base class' attribute when the child class _also_ had requirements, even if they are similar. It's not exactly that I wanted to be explicit, but rather that I'm trying to avoid entangling classes unnecessarily.

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/enzo/data_structures.py:987]: I think the `or` is deliberate, given how `EnzoDataset.__init__` handles `filename`, it really seems like there are 2 supported paths

File: yt/frontends/rockstar/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/rockstar/data_structures.py:118]: probably. especially because `read_cattrs` is able to raise `OSError` by itself. It seems harmless enough that I can include it in this PR.

File: yt/loaders.py
Comment (neutrinoceros) [reply yt/loaders.py:134]: good call, that must be a left over from a draft.

File: yt/frontends/rockstar/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/rockstar/data_structures.py]: (done)

**Comment by neutrinoceros** [comment]
FYI @chrishavlin I just forced pushed again to resolve a conflict. I don't intend to touch this branch anymore for now so feel free to play around with it

**Comment by neutrinoceros** [comment]
I would still be happy to see this in yt 4.2, but it can wait, and it doesn't seem worth blocking either, so I'll remove it from the milestone

**Comment by neutrinoceros** [comment]
I'll actually go one step further and add the "do not merge" label since it contains new deprecation warnings that need to reflect the first version they appeared in, and it doesn't look like it's going to be 4.2

**Comment by neutrinoceros** [comment]
I'm just making my best effort to avoid delaying the release forever with large review requests. My main motivation for pushing 4.2 these days is that I am uncomfortable with the gap that has grown between the 4.1.x backport branch and the dev branch. While bugs continue to be fixed regularly, cutting a bugfix release is becoming harder by the day, and I would like we keep releasing these fixes at a reasonable pace. In contrast, doing a feature release is reasonably easy *now*, but if we keep waiting for more PRs to go in we might end up painting ourselves in a corner.

Chris not being available to conclude his review ATM is also a reason why I'd rather postpone it :)

**Comment by neutrinoceros** [comment]
@chrishavlin gentle reminder :)

**Comment by chrishavlin** [comment]
Thanks for the ping! Should be able to get to this tomorrow!

File: yt/frontends/enzo/data_structures.py
Comment (chrishavlin) [inline yt/frontends/enzo/data_structures.py:987]: OK, odd quirk here! I just tried was loading `IsolatedGalaxy` in a fresh environment with a minimal yt install (so no h5py), and you get:

```
>>> ds = yt.load('IsolatedGalaxy/galaxy0030/galaxy0030')
<stdin>:1: UserWarning: This dataset appears to be of type EnzoDataset, but the following requirements are currently missing: h5py, libconf
Please verify your installation.
yt : [INFO     ] 2023-07-20 11:34:36,033 Parameters: current_time              = 0.0060000200028298
yt : [INFO     ] 2023-07-20 11:34:36,033 Parameters: domain_dimensions         = [32 32 32]
yt : [INFO     ] 2023-07-20 11:34:36,033 Parameters: domain_left_edge          = [0. 0. 0.]
yt : [INFO     ] 2023-07-20 11:34:36,034 Parameters: domain_right_edge         = [1. 1. 1.]
yt : [INFO     ] 2023-07-20 11:34:36,034 Parameters: cosmological_simulation   = 0
```
Based on the `_is_valid` here, the fact that the load appears to work (you actually get populated parameters!) would not be new to this PR , so it's actually quite helpful to have that `UserWarning` show up. But would it make sense to add an error here? 

Initially I was gonna suggest adding

```python
        if cls._missing_load_requirements():
            return False
```
like in the other updated `_is_valid` methods. But then that would print out the full list of h5py-dependent frontends that are possible. And here we actually know it's enzo.

**Review by chrishavlin — APPROVED** [review]
Ok! had a chance to test this out locally and it's great. I noted one odd quirk with enzo, but this is good to go after the conflicts are resolved IMO.

File: yt/frontends/enzo/data_structures.py
Comment (neutrinoceros) [reply yt/frontends/enzo/data_structures.py:987]: I think the paradigm has always been that `_is_valid` methods are not allowed to throw errors as it would give each frontend the power to break all others. Maybe I'm not understanding your suggestion ?

**Comment by neutrinoceros** [comment]
conflicts resolved

File: yt/frontends/enzo/data_structures.py
Comment (chrishavlin) [reply yt/frontends/enzo/data_structures.py:987]: Not suggesting to throw an error within the method. 

Just saying that you could refactor this is_valid to return False if the dependencies are missing, regardless of the filename convention check that it does now. That would result in a unidentified datatype error right away rather than a partially functional `ds`.

Comment (chrishavlin) [reply yt/frontends/enzo/data_structures.py:987]: I'm just on the fence on whether that it's better have a non functional ds but know that it's Enzo or an error right away and not know it's Enzo.

Comment (neutrinoceros) [reply yt/frontends/enzo/data_structures.py:987]: oh, I see. The behaviour you're pointing to is not intended ! I must have been testing this with `yt.load_sample("IsolatedGalaxy")` instead, which behaves differently. Your suggestion makes both loaders behave similarly, which is great, but we also loose the warning... I'll switch to draft while I try to think of a better solution.

Comment (neutrinoceros) [reply yt/frontends/enzo/data_structures.py:987]: ah, I see now that I replied in a hurry: as you already pointed, the behaviour of the Enzo frontend is weird on the main branch, and my patch only adds a warning to it (which is the point).
I'd prefer to keep this PR's scope limited to adding warnings and details to existing errors, rather than actively changing frontends' behaviour, just to keep things as tidy as possible. But I would merge your suggestion as a standalone patch ! WDYT ?

Comment (chrishavlin) [reply yt/frontends/enzo/data_structures.py:987]: Ya, that sounds good to me!

Comment (chrishavlin) [reply yt/frontends/enzo/data_structures.py:987]: Mostly wanted to bring up this behavior --- even without a follow-up patch, having the warning now is better than the behavior on main!

**Comment by neutrinoceros** [comment]
> this is good to go after the conflicts are resolved IMO.

@chrishavlin ? 👼🏻

**Review by chrishavlin — APPROVED** [review]

**Comment by neutrinoceros** [comment]
Ah, I forgot to update the deprecation messages (which is why I used the "do not merge" label). I'll do it now and hopefully we can merge then.

**Comment by neutrinoceros** [comment]
@chrishavlin sorry to ping you again. This is now ready to go !

**Review by chrishavlin — APPROVED** [review]

**Pull request merged**
